### PR TITLE
Sprite rendering

### PR DESCRIPTION
**Pull request by krupitskas**
Still need fix some code, remove another Nikita from authors

### CODE REVIEW

**Comment by Warezovvv** [comment]
Still need fix some code, remove another Nikita from authors

File: UnicornRender/video/include/unicorn/video/Material.hpp [hunk 1]
```diff
@@ -87,10 +87,54 @@ class Material
...
+     *
+     * @param x texture offset by x from 0.0 to 1.0
+     * @param y texture offset by y from 0.0 to 1.0
+     * @param width offset width from 0.0 to 1.0
+     * @param height offset height from 0.0 to 1.0
+    */
+    UNICORN_EXPORT void SetNormalizedTextureCoordinates(float x, float y, float width, float height);
+
+    /** @brief Drops previous set texture coordinates */
+    UNICORN_EXPORT void SetDefaultTextureCoodinates();
```
Comment (ilia-glushchenko) [inline UnicornRender/video/include/unicorn/video/Material.hpp]: I think the default values should be described directly here.

File: UnicornRender/video/include/unicorn/video/Material.hpp [hunk 2]
```diff
@@ -87,10 +87,54 @@ class Material
...
+
+    /**
+     * @brief Returns texture coordinates
+     *
+     * @note will generate warning if texture wasnt set
+     *
+     * @return texture coordinates
```
Comment (ilia-glushchenko) [inline UnicornRender/video/include/unicorn/video/Material.hpp]: The return format deserves description comment.

File: UnicornRender/video/include/unicorn/video/Material.hpp [hunk 3]
```diff
@@ -87,10 +87,54 @@ class Material
...
+    */
+    UNICORN_EXPORT glm::vec4 GetTextureCoordinates() const;
+
+    /**
+    * @brief Returns normalized texture coordinates
+    *
```
Comment (ilia-glushchenko) [inline UnicornRender/video/include/unicorn/video/Material.hpp]: The return format deserves a description comment.

File: UnicornRender/video/include/unicorn/video/vulkan/VkMesh.hpp [hunk 4]
```diff
@@ -81,6 +81,9 @@ class VkMesh
     /** @brief Returns constant reference to unicorn::Mesh */
     Mesh const& GetMesh() const;
 
+    /** @brief Returns sprite coordinates */
+    glm::vec4 GetSpriteCoordinates() const;
```
Comment (ilia-glushchenko) [inline UnicornRender/video/include/unicorn/video/vulkan/VkMesh.hpp]: I think the return format deserves some description.

**Review by ilia-glushchenko — CHANGES_REQUESTED** [review]

File: UnicornRender/video/include/unicorn/video/vulkan/VkMesh.hpp (hunk 4 above)
Comment (krupitskas) [reply UnicornRender/video/include/unicorn/video/vulkan/VkMesh.hpp]: Whoa, this line must not even exist)

File: UnicornRender/video/include/unicorn/video/Material.hpp [hunk 5]
```diff
@@ -87,10 +87,54 @@ class Material
...
     UNICORN_EXPORT glm::vec3 GetColor() const;
 
+    /**
+     * @brief Sets texture cordinates
+     *
+     * @note Can't set texture coordinates without texture.
+     *
+     * @param x texture offset by x in pixels
+     * @param y texture offset by y in pixels
+     * @param width offset width in pixels
```
Comment (MrDmitry) [inline UnicornRender/video/include/unicorn/video/Material.hpp]: offset width?

File: UnicornRender/video/include/unicorn/video/Material.hpp [hunk 6]
```diff
@@ -87,10 +87,54 @@ class Material
...
+     * @note Can't set texture coordinates without texture.
+     *
+     * @param x texture offset by x from 0.0 to 1.0
+     * @param y texture offset by y from 0.0 to 1.0
+     * @param width offset width from 0.0 to 1.0
+     * @param height offset height from 0.0 to 1.0
+    */
+    UNICORN_EXPORT void SetNormalizedTextureCoordinates(float x, float y, float width, float height);
+
+    /** @brief Drops previous set texture coordinates */
```
Comment (MrDmitry) [inline UnicornRender/video/include/unicorn/video/Material.hpp]: `Resets texture coordinates to default values`? And list values in the doxygen comment

File: UnicornRender/video/include/unicorn/video/Material.hpp [hunk 7]
```diff
@@ -87,10 +87,54 @@ class Material
...
+    UNICORN_EXPORT void SetDefaultTextureCoodinates();
+
+    /**
+     * @brief Returns texture coordinates
+     *
+     * @note will generate warning if texture wasnt set
```
Comment (MrDmitry) [inline UnicornRender/video/include/unicorn/video/Material.hpp]: If you mean `LOG_WARN` you don't need to mention it here. Also I'm not sold on the idea of `GetTextureCoordinates` checking anything to do about textures - it's just a vector of values, it doesn't matter if there is a texture bound or not

File: UnicornRender/video/source/Material.cpp [hunk 8]
```diff
@@ -104,5 +105,62 @@ glm::vec3 Material::GetColor() const
...
+        static_cast<float>(y) / m_albedo->Height(),
+        static_cast<float>(width) / m_albedo->Width(),
+        static_cast<float>(height) / m_albedo->Height() };
+
+    DataUpdated.emit();
+}
+
+void Material::SetNormalizedTextureCoordinates(float x, float y, float width, float height)
+{
+    if (!m_albedo->IsLoaded())
```
Comment (MrDmitry) [inline UnicornRender/video/source/Material.cpp]: why do you care if it's loaded or not? you're not accessing texture data at all here

File: UnicornRender/video/include/unicorn/video/Material.hpp [hunk 9]
```diff
@@ -87,10 +87,54 @@ class Material
...
+    *
+    * @return normalized texture coordinates
+    */
+    UNICORN_EXPORT glm::vec4 GetNormalizedTextureCoordinates() const;
+
     /** @brief Signal for material update notification */
     wink::signal<wink::slot<void()>> DataUpdated;
 protected:
     glm::vec3 m_color;
+    glm::vec4 m_spriteCoordinates;
```
Comment (MrDmitry) [inline UnicornRender/video/include/unicorn/video/Material.hpp]: I'm not sure if storing only normalized coordinates is the best approach.
Having explicit pixel coordinates helps with async textures where you don't yet know the actual size of the image but can already fill the sprite information. Also if you end up resizing your sprite atlas image (as in adding more frames/sprites) you'd want to use pixel coordinates, not the normalized ones. So in fact you only need normalized when you want to pass this information to the renderer, but not for the user.

I would think about storing both and having them work independently from texture (it doesn't make sense to lock the ability to set sprite coordinates when texture was not set - they should operate independently)
When texture is changed or when new coordinates are set (and texture is set) we can recalculate normalized ones and set them. I am not familiar with how sprite atlas images are typically used and if it's helpful to have normalized interface for setting texture coordinates. I feel like usually you'd use absolute values and normalized comes in handy _only_ when scaling your sprite resolutions which can also be mitigated with simply scaling the absolute values anyway... Not sure about normalized `set` interface, `get` makes sense for the render though.

File: UnicornRender/video/source/Material.cpp [hunk 10]
```diff
@@ -104,5 +105,62 @@ glm::vec3 Material::GetColor() const
...
+
+    m_spriteCoordinates = { x, y, width, height };
+
+    DataUpdated.emit();
+}
+
+
+void Material::SetDefaultTextureCoodinates()
+{
+    m_spriteCoordinates = { 0, 0, 1.0, 1.0 };
```
Comment (MrDmitry) [inline UnicornRender/video/source/Material.cpp]: I feel like this method is not needed - `SetNormalizedTextureCoordinates()` can be used instead

**Review by MrDmitry — CHANGES_REQUESTED** [review]
Is `UnicornRender/data/textures/sprite.png` a public domain image? Also change commit message, it should be brief and explain what the commit does.

**Review by ilia-glushchenko — APPROVED** [review]

File: UnicornRender/video/include/unicorn/video/Material.hpp [hunk 11]
```diff
@@ -87,10 +87,40 @@ class Material
     /** @brief Returns color */
     UNICORN_EXPORT glm::vec3 GetColor() const;
 
+    /**
+     * @brief Sets texture cordinates
+     *
+     * @note Can't set texture coordinates without texture.
```
Comment (MrDmitry) [inline UnicornRender/video/include/unicorn/video/Material.hpp]: outdated

File: UnicornRender/video/source/Material.cpp [hunk 12]
```diff
@@ -104,5 +115,35 @@ glm::vec3 Material::GetColor() const
...
+glm::vec4 Material::GetNormalizedTextureCoordinates() const
+{
+    return m_normSpriteCoords;
+}
+
+void Material::NormalizeTextureCoordinates()
+{
+    if (m_albedo->IsLoaded())
+    {
+        m_normSpriteCoords = { static_cast<float>(m_spriteCoords.x) / m_albedo->Width(),
```
Comment (MrDmitry) [inline UnicornRender/video/source/Material.cpp]: `m_spriteCoords` is a `glm::vec4`, why are you casting it to `float`? I feel like it makes more sense to cast `m_albedo->Width()` to `float`

**Review by MrDmitry — CHANGES_REQUESTED** [review]
Please change the sprite image to something from public domain or add appropriate attribution where it applies

**Review by ilia-glushchenko — APPROVED** [review]

File: UnicornRender/video/include/unicorn/video/Material.hpp [hunk 13]
```diff
@@ -87,10 +87,38 @@ class Material
...
+     * @param height sprite height in pixels
+     */
+    UNICORN_EXPORT void SetTextureCoordinates(int32_t x, int32_t y, int32_t width, int32_t height);
+
+    /**
+     * @brief Returns texture coordinates
+     *
+     * @return texture coordinates: {x, y, width, height}
+    */
+    UNICORN_EXPORT glm::vec4 GetTextureCoordinates() const;
```
Comment (MrDmitry) [inline UnicornRender/video/include/unicorn/video/Material.hpp]: Semantically coordinates is the wrong word - you return x, y coordinates and width, height dimensions. For coordinates you would've used `{x0, y0, x1, y1}`
So this is more a `GetSpriteArea`

**Review by MrDmitry — APPROVED** [review]

**Review by MrDmitry — APPROVED** [review]

**Comment by ilia-glushchenko** [comment]
Approve

[Approved with PullApprove](https://pullapprove.com/Godlike/Unicorn/pull-request/119/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=Godlike/Unicorn) [Approved with PullApprove](https://pullapprove.com/Godlike/Unicorn/pull-request/119/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=Godlike/Unicorn)

**Pull request merged**